# Springboard Data Science Career Track Unit 4 Challenge - Tier 3 Complete

## Objectives
Hey! Great job getting through those challenging DataCamp courses. You're learning a lot in a short span of time. 

In this notebook, you're going to apply the skills you've been learning, bridging the gap between the controlled environment of DataCamp and the *slightly* messier work that data scientists do with actual datasets!

Here’s the mystery we’re going to solve: ***which boroughs of London have seen the greatest increase in housing prices, on average, over the last two decades?***


A borough is just a fancy word for district. You may be familiar with the five boroughs of New York… well, there are 32 boroughs within Greater London [(here's some info for the curious)](https://en.wikipedia.org/wiki/London_boroughs). Some of them are more desirable areas to live in, and the data will reflect that with a greater rise in housing prices.

***This is the Tier 3 notebook, which means it's not filled in at all: we'll just give you the skeleton of a project, the brief and the data. It's up to you to play around with it and see what you can find out! Good luck! If you struggle, feel free to look at easier tiers for help; but try to dip in and out of them, as the more independent work you do, the better it is for your learning!***

This challenge will make use of only what you learned in the following DataCamp courses: 
- Prework courses (Introduction to Python for Data Science, Intermediate Python for Data Science)
- Data Types for Data Science
- Python Data Science Toolbox (Part One) 
- pandas Foundations
- Manipulating DataFrames with pandas
- Merging DataFrames with pandas

Of the tools, techniques and concepts in the above DataCamp courses, this challenge should require the application of the following: 
- **pandas**
    - **data ingestion and inspection** (pandas Foundations, Module One) 
    - **exploratory data analysis** (pandas Foundations, Module Two)
    - **tidying and cleaning** (Manipulating DataFrames with pandas, Module Three) 
    - **transforming DataFrames** (Manipulating DataFrames with pandas, Module One)
    - **subsetting DataFrames with lists** (Manipulating DataFrames with pandas, Module One) 
    - **filtering DataFrames** (Manipulating DataFrames with pandas, Module One) 
    - **grouping data** (Manipulating DataFrames with pandas, Module Four) 
    - **melting data** (Manipulating DataFrames with pandas, Module Three) 
    - **advanced indexing** (Manipulating DataFrames with pandas, Module Four) 
- **matplotlib** (Intermediate Python for Data Science, Module One)
- **fundamental data types** (Data Types for Data Science, Module One) 
- **dictionaries** (Intermediate Python for Data Science, Module Two)
- **handling dates and times** (Data Types for Data Science, Module Four)
- **function definition** (Python Data Science Toolbox - Part One, Module One)
- **default arguments, variable length, and scope** (Python Data Science Toolbox - Part One, Module Two) 
- **lambda functions and error handling** (Python Data Science Toolbox - Part One, Module Four) 

## The Data Science Pipeline

This is Tier Three, so we'll get you started. But after that, it's all in your hands! When you feel done with your investigations, look back over what you've accomplished, and prepare a quick presentation of your findings for the next mentor meeting. 

Data Science is magical. In this case study, you'll get to apply some complex machine learning algorithms. But as  [David Spiegelhalter](https://www.youtube.com/watch?v=oUs1uvsz0Ok) reminds us, there is no substitute for simply **taking a really, really good look at the data.** Sometimes, this is all we need to answer our question.

Data Science projects generally adhere to the four stages of Data Science Pipeline:
1. Sourcing and loading 
2. Cleaning, transforming, and visualizing 
3. Modeling 
4. Evaluating and concluding 


### 1. Sourcing and Loading 

Any Data Science project kicks off by importing  ***pandas***. The documentation of this wonderful library can be found [here](https://pandas.pydata.org/). As you've seen, pandas is conveniently connected to the [Numpy](http://www.numpy.org/) and [Matplotlib](https://matplotlib.org/) libraries. 

***Hint:*** This part of the data science pipeline will test those skills you acquired in the pandas Foundations course, Module One. 

#### 1.1. Importing Libraries

In [2]:
# Let's import the pandas, numpy libraries as pd, and np respectively. 
import pandas as pd 
import numpy as np

# Load the pyplot collection of functions from matplotlib, as plt 
import matplotlib.pyplot as plt

#### 1.2.  Loading the data
Your data comes from the [London Datastore](https://data.london.gov.uk/): a free, open-source data-sharing portal for London-oriented datasets. 

In [3]:
# First, make a variable called url_LondonHousePrices, and assign it the following link, enclosed in quotation-marks as a string:
# https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls
url_LondonHousePrices = "https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls"

# The dataset we're interested in contains the Average prices of the houses, and is actually on a particular sheet of the Excel file. 
# As a result, we need to specify the sheet name in the read_excel() method.
# Put this data into a variable called properties.  
properties = pd.read_excel(url_LondonHousePrices, sheet_name='Average price', index_col= None)

### 2. Cleaning, transforming, and visualizing
This second stage is arguably the most important part of any Data Science project. The first thing to do is take a proper look at the data. Cleaning forms the majority of this stage, and can be done both before or after Transformation.

The end goal of data cleaning is to have tidy data. When data is tidy: 

1. Each variable has a column.
2. Each observation forms a row.

Keep the end goal in mind as you move through this process, every step will take you closer. 



***Hint:*** This part of the data science pipeline should test those skills you acquired in: 
- Intermediate Python for data science, all modules.
- pandas Foundations, all modules. 
- Manipulating DataFrames with pandas, all modules.
- Data Types for Data Science, Module Four.
- Python Data Science Toolbox - Part One, all modules

**2.1. Exploring your data** 

Think about your pandas functions for checking out a dataframe. 

In [4]:
properties

Unnamed: 0 City of London Barking & Dagenham       Barnet       Bexley  \
0          NaT      E09000001          E09000002    E09000003    E09000004   
1   1995-01-01    91448.98487         50460.2266  93284.51832  64958.09036   
2   1995-02-01    82202.77314        51085.77983  93190.16963  64787.92069   
3   1995-03-01    79120.70256        51268.96956  92247.52435  64367.49344   
4   1995-04-01    77101.20804        53133.50526  90762.87492  64277.66881   
..         ...            ...                ...          ...          ...   
348 2023-12-01         804808             337725       581371       394474   
349 2024-01-01         798179             328359       581246       395468   
350 2024-02-01         785962             319809       573653       395439   
351 2024-03-01         755712             326650       582371       400146   
352 2024-04-01         778946             333425       573010       404557   

           Brent      Bromley       Camden      Croydon       Ealing  ...  \
0      E09000005    E09000006    E09000007    E09000008    E09000009  ...   
1    71306.56698  81671.47692  120932.8881  69158.16225  79885.89069  ...   
2    72022.26197  81657.55944  119508.8622  68951.09542  80897.06551  ...   
3    72015.76274  81449.31143  120282.2131  68712.44341  81379.86288  ...   
4    72965.63094  81124.41227   120097.899  68610.04641  82188.90498  ...   
..           ...          ...          ...          ...          ...  ...   
348       507592       493507       852047       388427       518410  ...   
349       520722       494361       824737       386875       517980  ...   
350       527606       493823       796724       393618       525523  ...   
351       510427       491226       738545       393334       520303  ...   
352       488462       485561       762130       398216       514465  ...   

      NORTH WEST YORKS & THE HUMBER EAST MIDLANDS WEST MIDLANDS  \
0      E12000002          E12000003     E12000004     E12000005   
1    43958.48001        44803.42878   45544.52227   48527.52339   
2    43925.42289        44528.80721   46051.57066   49341.29029   
3     44434.8681        45200.46775   45383.82395   49442.17973   
4     44267.7796        45614.34341   46124.23045   49455.93299   
..           ...                ...           ...           ...   
348       215263             204561        239962        246504   
349       213010             204703        237671        244903   
350       213455             205271        241168        243502   
351       215189             206052        240446        245345   
352       216714             203110        242132        246292   

    EAST OF ENGLAND       LONDON   SOUTH EAST   SOUTH WEST Unnamed: 47  \
0         E12000006    E12000007    E12000008    E12000009         NaN   
1        56701.5961  74435.76052  64018.87894   54705.1579         NaN   
2       56593.59475  72777.93709  63715.02399  54356.14843         NaN   
3       56171.18278  73896.84204  64113.60858  53583.07667         NaN   
4       56567.89582  74455.28754  64623.22395  54786.01938         NaN   
..              ...          ...          ...          ...         ...   
348          331316       511352       368405       308644         NaN   
349          334443       508972       371022       313714         NaN   
350          338133       506611       371224       314147         NaN   
351          337949       502703       369974       314378         NaN   
352          335448       501880       375067       317911         NaN   

         England  
0      E92000001  
1    53202.77128  
2     53096.1549  
3     53201.2843  
4     53590.8548  
..           ...  
348       296359  
349       296374  
350       297245  
351       297199  
352       298229  

[353 rows x 49 columns]

In [5]:
properties.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 353 entries, 0 to 352
Data columns (total 49 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Unnamed: 0            352 non-null    datetime64[ns]
 1   City of London        353 non-null    object        
 2   Barking & Dagenham    353 non-null    object        
 3   Barnet                353 non-null    object        
 4   Bexley                353 non-null    object        
 5   Brent                 353 non-null    object        
 6   Bromley               353 non-null    object        
 7   Camden                353 non-null    object        
 8   Croydon               353 non-null    object        
 9   Ealing                353 non-null    object        
 10  Enfield               353 non-null    object        
 11  Greenwich             353 non-null    object        
 12  Hackney               353 non-null    object        
 13  Hammersmith & Fulham

**2.2. Cleaning the data**

You might find you need to transpose your dataframe, check out what its row indexes are, and reset the index. You  also might find you need to assign the values of the first row to your column headings  . (Hint: recall the .columns feature of DataFrames, as well as the iloc[] method).

Don't be afraid to use StackOverflow for help  with this.

In [6]:
properties.dropna(how='all', axis=1, inplace=True)

In [7]:
properties = properties.loc[1:, 'Unnamed: 0': 'Westminster']

**2.3. Cleaning the data (part 2)**

You might we have to **rename** a couple columns. How do you do this? The clue's pretty bold...

In [8]:
properties.rename(columns={'Unnamed: 0': 'Date'}, inplace=True)

**2.4.Transforming the data**

Remember what Wes McKinney said about tidy data? 

You might need to **melt** your DataFrame here. 

In [9]:
properties = pd.melt(properties, id_vars = 'Date', var_name='Boroughs', value_name='Cost')

Remember to make sure your column data types are all correct. Average prices, for example, should be floating point numbers... 

In [10]:
properties_cost = properties.Cost.astype('float')

**2.5. Cleaning the data (part 3)**

Do we have an equal number of observations in the ID, Average Price, Month, and London Borough columns? Remember that there are only 32 London Boroughs. How many entries do you have in that column? 

Check out the contents of the London Borough column, and if you find null values, get rid of them however you see fit. 

In [11]:
properties.Boroughs.nunique()

33

**2.6. Visualizing the data**

To visualize the data, why not subset on a particular London Borough? Maybe do a line plot of Month against Average Price?

To limit the number of data points you have, you might want to extract the year from every month value your *Month* column. 

To this end, you *could* apply a ***lambda function***. Your logic could work as follows:
1. look through the `Month` column
2. extract the year from each individual value in that column 
3. store that corresponding year as separate column. 

Whether you go ahead with this is up to you. Just so long as you answer our initial brief: which boroughs of London have seen the greatest house price increase, on average, over the past two decades? 

In [12]:
properties['Date'] = properties['Date'].map(lambda p: p.year)

In [13]:
properties_gby = properties.groupby(['Boroughs', 'Date']).mean()

In [17]:
properties_gby = properties_gby.reset_index()

**3. Modeling**

Consider creating a function that will calculate a ratio of house prices, comparing the price of a house in 2018 to the price in 1998.

Consider calling this function create_price_ratio.

You'd want this function to:
1. Take a filter of dfg, specifically where this filter constrains the London_Borough, as an argument. For example, one admissible argument should be: dfg[dfg['London_Borough']=='Camden'].
2. Get the Average Price for that Borough, for the years 1998 and 2018.
4. Calculate the ratio of the Average Price for 1998 divided by the Average Price for 2018.
5. Return that ratio.

Once you've written this function, you ultimately want to use it to iterate through all the unique London_Boroughs and work out the ratio capturing the difference of house prices between 1998 and 2018.

Bear in mind: you don't have to write a function like this if you don't want to. If you can solve the brief otherwise, then great! 

***Hint***: This section should test the skills you acquired in:
- Python Data Science Toolbox - Part One, all modules

In [14]:
def create_price_ratio(j):
    yr_1998 = float(j['Cost'][j['Date']==1998])
    yr_2018= float(j['Cost'][j['Date']==2018])
    ratio = [yr_1998/yr_2018]
    return ratio

In [31]:
ratio_val = {} 
for v in properties_gby['Boroughs']:
   Boroughs_sel = properties_gby[properties_gby['Boroughs'] == v]
   ratio_val[v] = create_price_ratio(Boroughs_sel)


C:\Users\Justin\AppData\Local\Temp\ipykernel_27752\334717735.py:2: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  yr_1998 = float(j['Cost'][j['Date']==1998])
C:\Users\Justin\AppData\Local\Temp\ipykernel_27752\334717735.py:3: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  yr_2018= float(j['Cost'][j['Date']==2018])


In [34]:
sorted(ratio_val.items(), key=lambda x:x[1], reverse=True)

[('Hounslow', [0.25148317824115635]),
 ('Richmond upon Thames', [0.24967779731157863]),
 ('Harrow', [0.24635417785626296]),
 ('Bromley', [0.24421308489837312]),
 ('Sutton', [0.24280551426824518]),
 ('Hammersmith & Fulham', [0.24167443054605853]),
 ('Hillingdon', [0.23807975835429931]),
 ('Croydon', [0.23803288028014047]),
 ('Bexley', [0.2353507654063011]),
 ('Enfield', [0.23455064269011863]),
 ('Kingston upon Thames', [0.23416190234282552]),
 ('Ealing', [0.23194048191708755]),
 ('Havering', [0.23120155787014757]),
 ('Barnet', [0.22945274120785797]),
 ('Redbridge', [0.2236545053715767]),
 ('Tower Hamlets', [0.2161367227623553]),
 ('Merton', [0.21091380604361798]),
 ('Wandsworth', [0.2101851809159322]),
 ('Greenwich', [0.20995010893854218]),
 ('Islington', [0.20643891170300285]),
 ('Brent', [0.2043086864360114]),
 ('Barking & Dagenham', [0.20422256235393685]),
 ('Camden', [0.20261973503252542]),
 ('Lambeth', [0.20170435486140822]),
 ('Kensington & Chelsea', [0.19675491852791563]),
 ('Har

 The top 3 London boroughs that have seen the greatest average price increase
 over the past 20 years are 1.) Hounslow - 0.25148317824115635 2.) Richmond upon Thames - 0.24967779731157863
 3.) Harrow - 0.24635417785626296